### PR TITLE
Implement control logic for IPBlocksInfo

### PR DESCRIPTION
**Pull request by yanjunz97**
IPBlocksInfo will be
- periodically synchronized according to the existing VPCNetworkConfiguration CRs with a control loop
- updated with new created VPCNetworkConfiguration CR
- fully synchronized when a VPCNetworkConfiguration CR is deleted

Testing done:
- Verified IPBlocksInfo will be created according to the existing VPCNetworkConfiguration CRs.
Below is an example of IPBlocksInfo upon nsx-operator initialization after the first synchronization.
```
kind: IPBlocksInfo
metadata:
  creationTimestamp: "2024-08-27T11:58:59Z"
  generation: 1
  name: ip-blocks-info
  resourceVersion: "6267128"
  uid: 5b506e3d-79c6-4aeb-ae18-212349dd1ebd
externalIPCIDRs:
- 192.168.0.0/16
privateTGWIPCIDRs:
- 10.100.1.0/24
```
- Verified when a VPCNetworkConfiguration is created and deleted, the IPBlocksInfo CR will be updated as expected
Create a predefined VPCNetworkConfiguration CR as follows, which associated with the default project.
```
apiVersion: crd.nsx.vmware.com/v1alpha1
kind: VPCNetworkConfiguration
metadata:
  annotations:
    nsx.vmware.com/default: "true"
  name: vpc-network-config-with-pre-created-vpc
spec:
  nsxProject: /orgs/default/projects/project-quality
  vpc: /orgs/default/projects/project-quality/vpcs/773d3c87-3b88-43fa-ad6d-e20157404dd1
  defaultSubnetSize: 32
```
Got privateTGWIPCIDRs updated in the IPBlocksInfo CR
```
apiVersion: crd.nsx.vmware.com/v1alpha1
kind: IPBlocksInfo
metadata:
  creationTimestamp: "2024-08-27T11:58:59Z"
  generation: 2
  name: ip-blocks-info
  resourceVersion: "6281749"
  uid: 5b506e3d-79c6-4aeb-ae18-212349dd1ebd
externalIPCIDRs:
- 192.168.0.0/16
privateTGWIPCIDRs:
- 10.100.1.0/24
- 10.22.12.0/23
```
Delete the VPCNetworkConfiguration CR, observe the CIDRs associated with the deleted VPCNetworkConfiguration disappeared in IPBlocksInfo CR.

### CODE REVIEW

**Comment by wenqiq** [comment]
```
2024-08-27 04:16:18,571 INFO util.py:26 === RUN   TestInfraNetworkInfo
2024-08-27 04:16:19,584 INFO util.py:26 2024/08/27 04:16:19 kubectl get namespaces kube-system -o yaml | grep uid
2024-08-27 04:16:21,304 INFO util.py:26 2024/08/27 04:16:21 kubectl get networkinfos.crd.nsx.vmware.com kube-system -n kube-system -o yaml | grep uid
2024-08-27 04:16:21,795 INFO util.py:26     util.go:27:
2024-08-27 04:16:21,795 INFO util.py:26         	Error Trace:	/root/nsx-operator/test/e2e/util.go:27
2024-08-27 04:16:21,795 INFO util.py:26         	            				/root/nsx-operator/test/e2e/nsx_networkinfo_test.go:50
2024-08-27 04:16:21,795 INFO util.py:26         	            				/root/nsx-operator/test/e2e/nsx_networkinfo_test.go:90
2024-08-27 04:16:21,795 INFO util.py:26         	Error:      	Expected nil, but got: &errors.errorString{s:"error when running the following command `kubectl get networkinfos.crd.nsx.vmware.com kube-system -n kube-system -o yaml | grep uid` on master Node: <nil>, "}
2024-08-27 04:16:21,795 INFO util.py:26         	Test:       	TestInfraNetworkInfo
```
The e2e test failed, could you check it?

File: pkg/nsx/services/ipblocksinfo/ipblocksinfo.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+func (s *IPBlocksInfoService) getVPCConnectivityProfileByPath(path string) (*model.VpcConnectivityProfile, error) {
+	// An example of profile path: /orgs/{org-id}/projects/{project-id}/vpc-connectivity-profiles/{vpc-connectivity-profile-id}
+	pathSlice := strings.Split(path, "/")
+	if len(pathSlice) != 7 {
+		return nil, fmt.Errorf("failed to parse VPCConnectivityProfile path '%s'", path)
+	}
+	vpcConnectivityProfile, err := s.NSXClient.VPCConnectivityProfilesClient.Get(pathSlice[2], pathSlice[4], pathSlice[6])
+	return &vpcConnectivityProfile, err
+}
```
Comment (wenqiq) [inline pkg/nsx/services/ipblocksinfo/ipblocksinfo.go:301]: I think we can put this function in pkg/nsx/services/vpc/vpc.go:
https://github.com/vmware-tanzu/nsx-operator/pull/704

File: pkg/nsx/util/utils.go [hunk 2]
```diff
@@ -663,3 +665,22 @@ func UpdateRequestURL(reqUrl *url.URL, nsxtHost string, thumbprint string) {
 		reqUrl.Path = strings.Join(urls, "/")
 	}
 }
+
+func MergeArraysWithoutDuplicate(old []string, new []string) []string {
+	if old == nil || len(old) == 0 {
```
Comment (wenqiq) [inline pkg/nsx/util/utils.go:670]:
```suggestion
	if len(old) == 0 {
```

File: pkg/nsx/util/utils.go [hunk 3]
```diff
@@ -663,3 +665,22 @@ func UpdateRequestURL(reqUrl *url.URL, nsxtHost string, thumbprint string) {
...
+		return new
+	}
+	if new == nil || len(new) == 0 {
```
Comment (wenqiq) [inline pkg/nsx/util/utils.go:673]:
```suggestion
	if len(new) == 0 {
```

File: pkg/nsx/util/utils.go [hunk 4]
```diff
@@ -663,3 +665,22 @@ func UpdateRequestURL(reqUrl *url.URL, nsxtHost string, thumbprint string) {
...
+		return old
+	}
+	set := sets.New(old...)
+	set.Insert(new...)
+
+	return set.UnsortedList()
+}
+
+func CompareArraysWithoutOrder(old []string, new []string) bool {
+	oldSet := sets.New(old...)
```
Comment (wenqiq) [inline pkg/nsx/util/utils.go:683]:
```suggestion
	if len(old) != len(new) {
		return false
	}
	oldSet := sets.New(old...)
```

File: pkg/nsx/util/utils.go [hunk 5]
```diff
@@ -663,3 +665,22 @@ func UpdateRequestURL(reqUrl *url.URL, nsxtHost string, thumbprint string) {
 		reqUrl.Path = strings.Join(urls, "/")
 	}
 }
+
+func MergeArraysWithoutDuplicate(old []string, new []string) []string {
```
Comment (wenqiq) [inline pkg/nsx/util/utils.go:669]: I think we had better use another name, as `new` is a keyword in Golang.

File: pkg/controllers/networkinfo/vpcnetworkconfig_handler.go [hunk 6]
```diff
@@ -25,11 +25,12 @@ import (
...
+	Client              client.Client
+	vpcService          commontypes.VPCServiceProvider
+	ipBlocksInfoService commontypes.IPBlocksInfoServiceProvider
 }
 
-func (h *VPCNetworkConfigurationHandler) Create(_ context.Context, e event.CreateEvent, _ workqueue.RateLimitingInterface) {
+func (h *VPCNetworkConfigurationHandler) Create(ctx context.Context, e event.CreateEvent, _ workqueue.RateLimitingInterface) {
 	vpcConfigCR := e.Object.(*v1alpha1.VPCNetworkConfiguration)
 	vname := vpcConfigCR.GetName()
 	ninfo, _err := buildNetworkConfigInfo(*vpcConfigCR)
```
Comment (wenqiq) [inline pkg/controllers/networkinfo/vpcnetworkconfig_handler.go:36]: We can change `_err` here, and use `err`.

**Review by wenqiq — COMMENTED** [review]
Had better add a more detailed TestDone.

File: pkg/nsx/services/ipblocksinfo/ipblocksinfo.go (hunk 1 above)
Comment (yanjunz97) [reply pkg/nsx/services/ipblocksinfo/ipblocksinfo.go:320]: Updated, thanks!

File: pkg/nsx/util/utils.go (hunk 5 above)
Comment (yanjunz97) [reply pkg/nsx/util/utils.go]: Renamed to newArray, thanks!

File: pkg/controllers/networkinfo/vpcnetworkconfig_handler.go (hunk 6 above)
Comment (yanjunz97) [reply pkg/controllers/networkinfo/vpcnetworkconfig_handler.go]: Updated, thanks!

File: pkg/nsx/services/common/services.go [hunk 7]
```diff
@@ -1,9 +1,13 @@
 package common
 
 import (
+	"context"
+
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
```
Comment (wenqiq) [inline pkg/nsx/services/common/services.go:8]: The order of package imports typically follows these conventions to improve code readability and consistency:

Standard Library: Go's standard library packages.
Third-Party Libraries: third-party library packages.
Project-Specific Packages: custom packages within the project.

Comment (yanjunz97) [reply pkg/nsx/services/common/services.go:6]: Updated, thanks!

File: pkg/nsx/services/ipblocksinfo/ipblocksinfo.go [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+)
+
+var (
+	log                 = &logger.Log
+	ipBlocksInfoCRDName = "ip-blocks-info"
+	syncInterval        = 10 * time.Minute
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/nsx/services/ipblocksinfo/ipblocksinfo.go:27]: Maybe not as we currently do not have a specific field for Operator in NCP ConfigMap

**Review by ihgann — APPROVED** [review]
LGTM, but I'm not an expert on this operator. Will defer to other owners for final approval.

**Review by wenqiq — APPROVED** [review]
LGTM overall, since it's a big change we had better add some e2e test for this in another PR.

**Pull request merged**